### DISCUSSION TITLE
Adding a Custom Property View to a Theia GLSP Client

### DISCUSSION THREAD

**Discussion by Ginxss** [eclipse-glsp/glsp#860] [Q&A] [answered]
Hi,

I am trying to learn about how to use the Property View in Eclipse Theia. Unfortunately, I am already failing at the very first hurdle.

I took the current java-emf-theia project template, and placed a custom data service, content widget and widget provider next to the frontend-module in tasklist-theia, following [this tutorial](https://theia-ide.org/docs/property_view). The content of those files is identical to the example code.

I then registered them like this:
```typescript
export class TaskListTheiaFrontendModule extends GLSPTheiaFrontendModule {
    ...

    override configure(context: ContainerContext): void {
        context.bind(PropertyDataService).to(FileInfoPropertyDataService).inSingletonScope();
        context.bind(PropertyViewWidgetProvider).to(FileInfoPropertyViewWidgetProvider).inSingletonScope();
    }
}
```

However, the property view in Theia is still showing the default resource view. The `provideWidget` function of my widget provider is never getting called. But the `canHandle` function is getting called and returns the correct value.

Am I registering the components correctly?

**Comment by ndoschek** [comment, marked as answer]
Hi @Ginxss,
Thanks for your interest in GLSP and the Theia property view!

Your snippet looks good to me overall, however the super call is missing. Other than that and with the provided example code the property view should show your custom widget for the handled selection.

I can provide an example that listens to GLSP selections (as the linked tutorial is plain Theia documenation):
https://github.com/eclipse-glsp/glsp-examples/tree/property-view-example
This minimal example listens to GLSP selections and prints the ID of the selected GLSP diagram element.

HTH and best regards,
Nina

**Reply by Ginxss** [reply]
Thanks, that's very helpful!

**Reply by amiths89** [reply]
@ndoschek This is a great example. Just wondering how the configureClientActions() is done. Because I can't seem to find the function to override, It looks like it is not available in the glsp 2.0

**Reply by planger** [reply]
@amiths89 With 2.x it isn't necessary anymore to specify the client actions. As soon as there is a handler registered on the client, the server will send actions for the handled action types to the client.

**Reply by amiths89** [reply]
@planger I may have not registered the Action ID. Is this done differently now?

**Reply by amiths89** [reply]
```
@injectable()
export class TasklistGLSPTheiaDiagramServer extends GLSPTheiaDiagramServer {
    override initialize(registry: ActionHandlerRegistry): void {
        super.initialize(registry);
        registry.register(REQUEST_ELEMENT_DATA_ACTION_ID, this);
    }
}

```

**Reply by planger** [reply]
You'd just need to [register an action handler](https://eclipse.dev/glsp/documentation/actionhandler/#implementing-an-action-handler-glsp-client) on the client. That should be enough for it to be sent to the server and then forwarded to the client if the action is triggered.

**Reply by amiths89** [reply]
Thanks Philip. I tried this but i still seem to be having a circular dependency issue in one of the 'toDynamicValue' bindings when binding GlspSelectionDataService in TaskListDiagramConfiguration
`container.bind(GlspSelectionDataService).to(CustomGlspSelectionDataService);`

Actually if there's any recent Theia Properties selection example that could help.

**Reply by planger** [reply]
Sorry to hear that! I'm not sure whether there is a recent example for the Theia Properties panel in combination with GLSP publicly available. Probably it is best if you open a new discussion with your concrete problem, current code and error message. This way it is probably easier to help. Thank you!

**Reply by amiths89** [reply]
Sure I ll open a new discussion.

**Reply by rsoika** [reply]
Hi @amiths89 you can also take a look into our project - [Open-BPMN](https://www.open-bpmn.org). We implemented a Property Panel as an UI Extension. I know this is not exactly what you are searching for. But the reason why we implemented a Property Panel as an UI Extension was that this works on all platforms because the panel is part of the diagram plane. So no extra panel/view need to be implemented. But of course also this approach has some complexity. And we use a lot of custom Action Events too ;-)
You can take a look at our [online demo](https://open-bpmn.imixs.org/#/usr/src/app/open-bpmn.glsp-client/workspace/demo). Just double click on an element and the panel will pop up. We are using JsonForms to implement the different element properties.

**Reply by amiths89** [reply]
Thanks @rsoika for the demo link. Actually I have already been looking at this code. But since i was just starting the Theia Property View seemed like a more easier place to start with.
